### ISSUE TITLE
Issue with Mean, Argmax

### ISSUE THREAD

**Issue by shijianjian** [tf-coreml/tf-coreml#177]
I am facing issues with Mean, Argmax ops.

Since I am processing 3D point clouds than regular 2D images. Hence the input is not essentially (S, H, W, C). My input is (S, points, C). Due to the fewer dimensions, those errors comes out:

With Mean op:
Mean axis case not handled currently. Input shape = [1, 256], output shape = [1, 256], axis_ind = [0]

With Argmax op:
Mean axis case not handled currently. Input shape = (1, 4096, 13), output shape = (1, 4096), axis_ind = 'tf/argmax/dimensions:0'.

I have tweaked the code accordingly to make it work while I am not sure am I going on the right path.

I assume that in Mean op, if the input axis_ind = 0, the mean operation shall on the 'S' axis.
In Argmax op, we just need to follow the axis where it pointed to (CHW), as long as it is not 'S' (coremltools does not support it).

I will append a PR soon.

**Comment by aseemw** [comment]
Please see my comments in the PR #178. 

The argmax case will be difficult to handle without potentially breaking some other models. 
Can you share the model you are using? There might be other ways of dealing with it.

**Comment by shijianjian** [comment]
@aseemw Thanks for your input of the CoreML logics.

The model I am using is from [PointNet](https://github.com/charlesq34/pointnet), I cannot post the whole model here, I will give the gist related to this particular issue.

In our case, the sample input is (10, 4096, 9) which represents 10 batches of pointclouds, 4096 points and 9 channels in each pointcloud.

This is the main model:
```python
def get_model(point_cloud, keep_rate, bn_decay=None):
    """ ConvNet baseline, input is BxNx3 gray image """
    batch_size = point_cloud.get_shape()[0].value
    num_point = point_cloud.get_shape()[1].value

    input_image = tf.expand_dims(point_cloud, -1)
    # CONV
    net = tf_util.conv2d(input_image, 64, [1,9], padding='VALID', stride=[1,1],
                         bn=True, scope='conv1', bn_decay=bn_decay)
    net = tf_util.conv2d(net, 64, [1,1], padding='VALID', stride=[1,1],
                         bn=True, scope='conv2', bn_decay=bn_decay)
    net = tf_util.conv2d(net, 64, [1,1], padding='VALID', stride=[1,1],
                         bn=True, scope='conv3', bn_decay=bn_decay)
    net = tf_util.conv2d(net, 128, [1,1], padding='VALID', stride=[1,1],
                         bn=True, scope='conv4', bn_decay=bn_decay)
    points_feat1 = tf_util.conv2d(net, 1024, [1,1], padding='VALID', stride=[1,1],
                         bn=True, scope='conv5', bn_decay=bn_decay)
    # MAX
    pc_feat1 = tf_util.max_pool2d(points_feat1, [num_point,1], padding='VALID', scope='maxpool1')
    # FC
    pc_feat1 = tf.reshape(pc_feat1, [batch_size, -1])
    pc_feat1 = tf_util.fully_connected(pc_feat1, 256, bn=True, scope='fc1', bn_decay=bn_decay)
    pc_feat1 = tf_util.fully_connected(pc_feat1, 128, bn=True, scope='fc2', bn_decay=bn_decay)
    print(pc_feat1)
   
    # CONCAT 
    pc_feat1_expand = tf.concat([tf.reshape(pc_feat1, [batch_size, 1, 1, -1]) for _ in range(num_point)], axis=1)
    # pc_feat1_expand = tf.tile(tf.reshape(pc_feat1, [batch_size, 1, 1, -1]), [1, num_point, 1, 1])
    points_feat1_concat = tf.concat(axis=3, values=[points_feat1, pc_feat1_expand])
    
    # CONV 
    net = tf_util.conv2d(points_feat1_concat, 512, [1,1], padding='VALID', stride=[1,1],
                         bn=True, scope='conv6')
    net = tf_util.conv2d(net, 256, [1,1], padding='VALID', stride=[1,1],
                         bn=True, scope='conv7')
    net = tf_util.dropout(net, keep_prob=keep_rate, scope='dp1')
    net = tf_util.conv2d(net, 13, [1,1], padding='VALID', stride=[1,1],
                         activation_fn=None, scope='conv8')
    net = tf.squeeze(net, [2])

    return net
```
While the mean happens when using batch normalization, particularly in **tf.nn.moments()** function to obtain the mean and variance.
```python

def batch_norm_template(inputs, scope, moments_dims, bn_decay):
  """ Batch normalization on convolutional maps and beyond...
  Ref.: http://stackoverflow.com/questions/33949786/how-could-i-use-batch-normalization-in-tensorflow
  
  Args:
      inputs:        Tensor, k-D input ... x C could be BC or BHWC or BDHWC
      scope:         string, variable scope
      moments_dims:  a list of ints, indicating dimensions for moments calculation
      bn_decay:      float or float tensor variable, controling moving average weight
  Return:
      normed:        batch-normalized maps
  """
  with tf.variable_scope(scope) as sc:
    num_channels = inputs.get_shape()[-1].value
    beta = tf.Variable(tf.constant(0.0, shape=[num_channels]),
                       name='beta', trainable=True)
    gamma = tf.Variable(tf.constant(1.0, shape=[num_channels]),
                        name='gamma', trainable=True)
    batch_mean, batch_var = tf.nn.moments(inputs, moments_dims, name='moments')
    decay = bn_decay if bn_decay is not None else 0.9
    ema = tf.train.ExponentialMovingAverage(decay=decay)

    # Operator that maintains moving averages of variables.
    ema_apply_op = ema.apply([batch_mean, batch_var])
    
    # Update moving average and return current batch's avg and var.
    def mean_var_with_update():
      with tf.control_dependencies([ema_apply_op]):
        return tf.identity(batch_mean), tf.identity(batch_var)
    
    # ema.average returns the Variable holding the average of var.
    mean, var = mean_var_with_update()

    normed = tf.nn.batch_normalization(inputs, mean, var, beta, gamma, 1e-3)
  return normed
```
ArgMax happens when obtain the results. exp. we have output of (1, 4096, 13) which means there are 4096 points and each points are one-hot encoded with a specific label out of 13 classes. We would use ArgMax to obtain the final results. Code like below:
```python
correct = tf.equal(tf.argmax(pred, 2), tf.to_int64(labels_pl))
```
Let's see what we can do to solve the problem. I do appreciate your time on this.

**Comment by shijianjian** [comment]
By the way, I have tried to compile the TF model to CoreML model using the code I implemented. I seems like all goes well. It did produce a .mlmodel while the process hangs after
```bash
...
8615/8643: Converting op name: output/ArgMax ( type:  ArgMax )
8616/8643: Converting op name: output/Squeeze ( type:  Squeeze )

 Core ML model generated. Saved at location: ./model.mlmodel

Core ML input(s):
 [name: "input__Placeholder_1__0"
type {
  multiArrayType {
    shape: 1
    shape: 1
    shape: 1
    dataType: DOUBLE
  }
}
, name: "input__Placeholder__0"
type {
  multiArrayType {
    shape: 1
    shape: 4096
    shape: 9
    dataType: DOUBLE
  }
}
]
Core ML output(s):
 [name: "output__Squeeze__0"
type {
  multiArrayType {
    shape: 4096
    shape: 1
    shape: 1
    dataType: DOUBLE
  }
}
]
```
I can't even stop the process with **ctrl+c**

I tried with the latest commit again
```bash
...
8615/8643: Analysing op name: output/ArgMax ( type:  ArgMax )
8616/8643: Analysing op name: output/Squeeze ( type:  Squeeze )
8617/8643: Analysing op name: conv7/bn/ExponentialMovingAverage/AssignMovingAvg_1/sub_1 ( type:  Sub )

 Core ML model generated. Saved at location: ./model.mlmodel

Core ML input(s):
 [name: "input__Placeholder_1__0"
type {
  multiArrayType {
    shape: 1
    shape: 1
    shape: 1
    dataType: DOUBLE
  }
}
, name: "input__Placeholder__0"
type {
  multiArrayType {
    shape: 1
    shape: 4096
    shape: 9
    dataType: DOUBLE
  }
}
]
Core ML output(s):
 [name: "output__Squeeze__0"
type {
  multiArrayType {
    shape: 4096
    shape: 1
    shape: 1
    dataType: DOUBLE
  }
}
]
```
By your implementation, there are one more step forward which is (**8617/8643**) while I am not sure but it seems like a few more operations are skipped without notice.

**Comment by aseemw** [comment]
The conversion should stop after the "output_fearure_name" is reached. So it doesn't need to go till the last op. 

Its surprising that the program is hanging, especially after it has saved out the .mlmodel file, after that it should just exit while returning the MLModel. 

So with the latest master + your tweak for the argmax op, is the mlmodel produced valid? That is, does it give same predictions as the TF model?

**Comment by shijianjian** [comment]
I think none of our works works on my model.

The hanging point comes from the last line of convert(...) function
```python
  ...
  return MLModel(spec)
```
The coremltools itself could not read the converted model file. I think that is why it hangs, it may be some unexpected errors inside. I tried to debug this while there is an **libcoremlpython.so** package in coremltools project that I cannot access.

When I import those .mlmodel models into XCode, it gives
![image](https://user-images.githubusercontent.com/15955486/40168962-c5f69e36-59bc-11e8-946b-22e16d1452ed.png)

This is my model:
[model.zip](https://github.com/tf-coreml/tf-coreml/files/2012867/model.zip)

And code for convert:
```python
frozen_model_file = os.path.abspath('./saved_model.pbtxt')
predicted_probabilities_output = ['ceiling', 'floor', 'wall', 'beam', 'column', 'window', 'door', 'table', 'chair', 'sofa', 'bookcase', 'board', 'clutter']
input_tensor_shapes = {"input/Placeholder:0": [1,4096, 9], "input/Placeholder_1:0": []} # batch size is 1
# Output CoreML model path
coreml_model_file = './pointnet.mlmodel'
output_tensor_names = ['output/Squeeze:0']

tfcoreml.convert(
            tf_model_path=frozen_model_file, 
            mlmodel_path=coreml_model_file, 
            input_name_shape_dict=input_tensor_shapes,
            output_feature_names=output_tensor_names,
            predicted_probabilities_output=predicted_probabilities_output)
```

**Comment by aseemw** [comment]
A few comments:
- `saved_model.pbtxt` is a binary file, so it should be renamed to `saved_model.pb` for clarity
- The output of the TF graph `output/Squeeze:0` is a 4096 length vector. So `predicted_probabilities_output` must be of the same length. 
- `input/Placeholder_1:0` is a dropout related input. Ideally it should be removed from the frozen graph, but I'm not sure how to do that. However, its a scalar so during conversion, its specified shape should be (1). [See discussion in issue #169  and issue #168]

**Comment by aseemw** [comment]
Looking at the text summary (/utils/inspect_pb.py) of the frozen graph, I see 

OPS counts:
RealDiv : 1
Slice : 1
ExpandDims : 1
RandomUniform : 1
Floor : 1
ArgMax : 1
Softmax : 1
Pack : 1
MaxPool : 1
Placeholder : 2
MatMul : 2
ConcatV2 : 3
Conv2D : 8
Relu : 9
StopGradient : 9
Rsqrt : 9
SquaredDifference : 9
NoOp : 9
BiasAdd : 10
Mean : 18
Add : 20
Squeeze : 20
Mul : 47
Sub : 65
Identity : 74
Reshape : 4099
Const : 4221

There are too many reshapes, and the concat op is massive (concatenating abt 4000 tensors). That makes the graph evaluation and conversion super slow. 
I am not sure of the architecture details, but IMO, the graph should be simplified first before converting. Try to get rid of the massive concat and the the reshapes.

**Comment by aseemw** [comment]
The input `input/Placeholder:0` has shape `(1, 4096, 9)`, but the the first operation it undergoes is `expand_dims` and it becomes 4-rank `(1, 4096, 9, 1)`. Since the tfcoreml converter is generally much more stable when dealing with rank-4 tensors its better to re-freeze the graph such that its input is 4-rank. 

Here is how you can do it (copying a bunch of code from one of the jupyter-notebook examples):

```
import tensorflow as tf
from tensorflow.core.framework import graph_pb2
import numpy as np

tf_model_path = './saved_model.pb'


tf.reset_default_graph()
original_gdef = tf.GraphDef()

with open(tf_model_path, 'rb') as f:
    serialized = f.read()
original_gdef.ParseFromString(serialized)

with tf.Graph().as_default() as g:
    tf.import_graph_def(original_gdef, name='')
    
    ops = g.get_operations()
    
    for i in range(10):
            print('\n\nop id {} : op type: "{}"'.format(str(i), ops[i].type));
            print('input(s):'),
            for x in ops[i].inputs:
                print("name = {}, shape: {}, ".format(x.name, x.get_shape())),
            print('\noutput(s):'),
            for x in ops[i].outputs:
                print("name = {}, shape: {},".format(x.name, x.get_shape())),
                
X = np.random.rand(1, 4096, 9)
                
with tf.Session(graph = g) as sess: 
    out1 = sess.run('output/Squeeze:0',feed_dict={'input/Placeholder:0': X, 'input/Placeholder_1:0': np.ones((1))})
        
print ""
print "final output shape: ", out1.shape
        

from tensorflow.python.tools import strip_unused_lib
from tensorflow.python.framework import dtypes
from tensorflow.python.platform import gfile
input_node_names = ['ExpandDims']
output_node_names = ['output/Squeeze']
gdef = strip_unused_lib.strip_unused(
        input_graph_def = original_gdef,
        input_node_names = input_node_names,
        output_node_names = output_node_names,
        placeholder_type_enum = dtypes.float32.as_datatype_enum)

frozen_model_file = './saved_model_refrozen.pb'
with gfile.GFile(frozen_model_file, "wb") as f:
    f.write(gdef.SerializeToString())
    
tf.reset_default_graph()
new_gdef = tf.GraphDef()

with open(frozen_model_file, 'rb') as f:
    serialized = f.read()
new_gdef.ParseFromString(serialized)

with tf.Graph().as_default() as g:
    tf.import_graph_def(new_gdef, name='')
    
    ops = g.get_operations()
    
    for i in range(10):
            print('\n\nop id {} : op type: "{}"'.format(str(i), ops[i].type));
            print('input(s):'),
            for x in ops[i].inputs:
                print("name = {}, shape: {}, ".format(x.name, x.get_shape())),
            print('\noutput(s):'),
            for x in ops[i].outputs:
                print("name = {}, shape: {},".format(x.name, x.get_shape())),
                
X = np.reshape(X, (1, 4096, 9, 1))

with tf.Session(graph = g) as sess:
    out2 = sess.run('output/Squeeze:0',feed_dict={'ExpandDims:0': X, 'input/Placeholder_1:0': np.ones((1))})

print ""
print "final output shape: ", out2.shape
print "difference between the output of the two TF graphs: ", np.sum(np.abs(out1-out2))
```

Once that is done, 
conversion should work (but its super slow) and the evaluation is insanely slow because of the concat/reshape issue.  

```
output = 'output/Squeeze:0'
coreml_model_file = './pointnet.mlmodel'
output_tensor_names = [output]
input_tensor_shapes = {'ExpandDims:0':(1,4096,9,1) , 'input/Placeholder_1:0': [1]}

print("Starting conversion.....")

cm = tfcoreml.convert(
            tf_model_path='saved_model_refrozen.pb',
            mlmodel_path=coreml_model_file,
            input_name_shape_dict=input_tensor_shapes,
            output_feature_names=output_tensor_names)
```

**Comment by shijianjian** [comment]
@aseemw Thanks for your comments. They are pretty helpful.

Originally, the model used **tf.tile()** operation which is not supported by **tf-coreml**. Hence I changed it to **tf.concat()** for converting. 

I realized where the problem lies by going through your comments. So I manually implemented the **tf.tile()** method.

Here is the numpy implementation.

I put my numpy implementation in this [gist](https://gist.github.com/shijianjian/36c33e903fca79f070aa11bdbb033706).

---
When I tried to implement tile op for tf-coreml, I need some particular operations while I found the expand_dims operation has been annotated as <i>dummy operation</i>. How can I call it in the middle of conversion?

```python
# dummy for CoreML
  'ExpandDims' : _layers.skip,
```
---

For people coming later. You may need the tensorflow implemtation.

```python
def tile(inputs, multiplier):
  """ Implementation on tf.tile layer

  Args:
    inputs: tensor
    multiplier: list of multiplier along each dimensions

  Returns:
    tensor variable
  """
  if not isinstance(multiplier, (list, np.ndarray)):
      raise ValueError("multiplier must be a List or numpy array")
  
  multiplier = np.array(multiplier)
      
  input_shape = inputs.get_shape().as_list()
  expand_shape = len(multiplier) - len(input_shape)
  
  # Make array and multiplyiers to be the same sized
  if expand_shape > 0:
    init_mul = np.concatenate([np.ones(expand_shape, dtype=np.int32), input_shape], axis=0)
    inputs = tf.multiply(inputs, np.ones(init_mul))
  else:
    multiplier = np.concatenate([np.ones(-expand_shape, dtype=np.int32), multiplier], axis=0)
  
  for i, dim_mul in enumerate(multiplier):
    _dim_mul = np.insert(input_shape, i, dim_mul)
    intermedia = tf.expand_dims(inputs, i)
    intermedia = tf.multiply(intermedia, np.ones(_dim_mul))
    
    _squeeze_mul = input_shape
    _squeeze_mul[i] = dim_mul*input_shape[i]
    _squeeze_mul = np.insert(_squeeze_mul, i, -1)

    intermedia = tf.reshape(intermedia, _squeeze_mul)
    inputs = tf.squeeze(intermedia, i)

  return inputs
```

**Comment by aseemw** [comment]
PR #179 might be useful in this case. You should check that out.

**Comment by Mstronach** [comment]
These ops have been included in Core ML 3, please try it out and let us know if you are still seeing the issue. Thanks!
